File: problems/2-medium/19-Remove-Nth-Node-From-End-Of-List/solution.java
public class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {

        ListNode slow = head;
        ListNode fast = head;
        ListNode prev = null;

        for (int i = 1; i < n; i++) {
            fast = fast.next;
        }

        while (fast.next != null) {
            prev = slow;
            slow = slow.next;
            fast = fast.next;
        }
           
        if (prev == null) {
            head = head.next;
        } else {
            prev.next = slow.next;
        }

        return head;
        
    }
}

File: problems/2-medium/19-Remove-Nth-Node-From-End-Of-List/solution.cpp
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};


ListNode* removeNthFromEnd(ListNode* head, int n) {
    ListNode* fast = head;
    for (int i = 0; i < n; i++) {
        fast = fast->next;
    }

    ListNode* slow = head;
    ListNode* prev = nullptr;
    while (fast) {
        fast = fast->next;
        prev = slow;
        slow = slow->next;
    }

    if (prev == nullptr) return head->next;

    prev->next = prev->next->next;
    return head;

}
